include/envoy/server/lifecycle_notifier.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <functional>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Server {
+
+class ServerLifecycleNotifier {
+public:
+  virtual ~ServerLifecycleNotifier() = default;
+
+  /**
+   * Stages of the envoy server instance lifecycle.
+   */
+  enum class Stage {
+    /**
+     * The server instance main thread has entered the dispatcher loop.
+     */
+    Startup,
+
+    /**
+     * The server instance is being shutdown and the dispatcher is about the exit.
+     */
+    ShutdownExit
+  };
```

Comment:
This API restricts us from distinguishing between normal and abnormal shutdowns in the future. Do you think it's useful to build that capability in? It's unclear from the PR description what the motivation is for this change so I can't answer that without more context.

Reply:
We can easily add additional values to the Stages enum to represent certain failure cases. Also it should be easy to extend the API to pass a parameter in the StageCallback function so that we can pass things like an error code or other state to the listeners. I'd rather wait till we have a more concrete use case for that though and keep things simple for now.

The bug for this (#2802) has some more context about the motivation for this change. In my particular use case I have a filter that needs to be given a chance to flush cached state when envoy is shutting down.

Reply:
Sounds good to me, thanks for expanding the comment.